src/solve/SimulationParameters.cpp

```diff
@@ -32,35 +32,143 @@
 
 #include "State.h"
 
-std::vector<double> get_double_array(const nlohmann::json& data,
-                                     const std::string_view& key) {
-  std::vector<double> vector;
-  if (!data[key].is_array()) {
-    return {data[key]};
+bool get_param_scalar(const nlohmann::json& data, const std::string& name,
+                      const InputParameter& param, double& val) {
+  if (data.contains(name)) {
+    val = data[name];
+  } else {
+    if (param.is_optional) {
+      val = param.default_val;
+    } else {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool get_param_vector(const nlohmann::json& data, const std::string& name,
+                      const InputParameter& param, std::vector<double>& val) {
+  if (data.contains(name)) {
+    val = data[name].get<std::vector<double>>();
+  } else {
+    if (param.is_optional) {
+      val = {param.default_val};
+    } else {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool has_parameter(
+    const std::vector<std::pair<std::string, InputParameter>>& params,
+    const std::string& name) {
+  for (const auto& pair : params) {
+    if (pair.first == name) {
+      return true;
+    }
   }
-  return data[key].get<std::vector<double>>();
+  return false;
 }
 
-std::vector<double> get_double_array(const nlohmann::json& data,
-                                     const std::string_view& key,
-                                     const std::vector<double>& default_value) {
-  if (!data.contains(key)) {
-    return default_value;
+int generate_block(Model& model, const nlohmann::json& config,
+                   const std::string& block_name, const std::string_view& name,
```

Comment:
Should `block_name` be named `block_type` instead?

Reply:
Yes!